src/network/protocol/messages/block.zig

```diff
@@ -0,0 +1,201 @@
+
+        try self.txn_count.encodeToWriter(w);
+
+        for (self.txns.items) |txn| {
+            try txn.serializeToWriter(w);
+        }
+    }
+
+    /// Serialize a message as bytes and write them to the buffer.
+    ///
+    /// buffer.len must be >= than self.hintSerializedLen()
+    pub fn serializeToSlice(self: *const Self, buffer: []u8) !void {
+        var fbs = std.io.fixedBufferStream(buffer);
+        try self.serializeToWriter(fbs.writer());
+    }
+
+    /// Serialize a message as bytes and return them.
+    pub fn serialize(self: *const BlockMessage, allocator: std.mem.Allocator) ![]u8 {
+        const serialized_len = self.hintSerializedLen();
+
+        const ret = try allocator.alloc(u8, serialized_len);
+        errdefer allocator.free(ret);
+
+        try self.serializeToSlice(ret);
+
+        return ret;
+    }
+
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !BlockMessage {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(r), "readInt")) @compileError("Expects r to have fn 'readInt'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readNoEof")) @compileError("Expects r to have fn 'readNoEof'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readAll")) @compileError("Expects r to have fn 'readAll'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readByte")) @compileError("Expects r to have fn 'readByte'.");
+        }
+
+        var bm: Self = undefined;
+
+        bm.block_header.version = try r.readInt(i32, .little);
+
+        const prev_block_raw_bytes = try read_bytes_exact(allocator, r, 32);
+        defer allocator.free(prev_block_raw_bytes);
+        @memcpy(&bm.block_header.prev_block, prev_block_raw_bytes);
+
+        const merkle_root_raw_bytes = try read_bytes_exact(allocator, r, 32);
+        defer allocator.free(merkle_root_raw_bytes);
+        @memcpy(&bm.block_header.merkle_root, merkle_root_raw_bytes);
+
+        bm.block_header.timestamp = try r.readInt(i32, .little);
+        bm.block_header.nbits = try r.readInt(i32, .little);
+        bm.block_header.nonce = try r.readInt(i32, .little);
+        bm.txn_count = try CompactSizeUint.decodeReader(r);
+
+        bm.txns = std.ArrayList(Transaction).init(allocator);
+        errdefer bm.txns.deinit();
+
+        const txns_count_u32 = bm.txn_count.value();
+        while (bm.txns.items.len < txns_count_u32) {
+            try bm.txns.append(try Transaction.deserializeReader(allocator, r));
+        }
+
+        return bm;
+    }
+
+    /// Deserialize bytes into a `VersionMessage`
+    pub fn deserializeSlice(allocator: std.mem.Allocator, bytes: []const u8) !Self {
+        var fbs = std.io.fixedBufferStream(bytes);
+        return try Self.deserializeReader(allocator, fbs.reader());
+    }
+
+    pub fn hintSerializedLen(self: BlockMessage) usize {
+        const header_length = @sizeOf(BlockHeader);
+        const txs_number_length = @sizeOf(CompactSizeUint);
```

Comment:
nope. It's `CompactSizeUint.new(self.txns.items.len).hint_encoded_len()`.

It's the size of the future serialized item, not the size of the struct in-memory

Reply:
Oh, I didn't know CompactSizeUint could have different length, as I saw a u64 as the inner value